### PR TITLE
MLIR SROA

### PR DESCRIPTION
**Pull request by Moxinilian**
Implementation of a simple form of generic SROA in MLIR.

### CODE REVIEW

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.h [hunk 1]
```diff
@@ -26,6 +26,12 @@ struct MemorySlot {
   Location location;
 };
 
+struct DestructibleMemorySlot : public MemorySlot {
+  /// Maps from an index within the memory slot the type of the pointer that
+  /// will be generated to access the element directly.
+  DenseMap<Attribute, Type> elementsPtrs;
```
Comment (gysit) [inline mlir/include/mlir/Interfaces/MemorySlotInterfaces.h:32]: nit: elementPtrs?

File: mlir/lib/Transforms/Mem2Reg.cpp [hunk 2]
```diff
@@ -472,9 +499,11 @@ LogicalResult mlir::tryToPromoteMemorySlots(
       if (slot.ptr.use_empty())
         continue;
 
-      SlotPromoter promoter(slot, allocator, builder, dominance);
-      if (succeeded(promoter.prepareSlotPromotion()))
-        toPromote.emplace_back(std::move(promoter));
+      SlotPromotionAnalyzer analyzer(slot, dominance);
+      Optional<SlotPromotionMetadata> metadata = analyzer.computeMetadata();
```
Comment (gysit) [inline mlir/lib/Transforms/Mem2Reg.cpp:503]: metadata is maybe a bit overused for a project that sits in the llvm-project directory. 

SlotPromotionAnalysis, SlotPromotionInfo (I know terrible), I know you have a better idea :)

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.h (hunk 1 above)
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.h:32]: subelementPointers?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+  Optional<Type> maybeSelectedType = gep.getElemType();
+  if (!maybeSelectedType)
+    return {};
+  Type selectedType = maybeSelectedType.value();
+  Attribute firstLevelIndex;
+  for (const auto &index : gep.getIndices()) {
+    IntegerAttr indexInt = index.dyn_cast<IntegerAttr>();
+    if (!indexInt)
+      return {};
+    if (isCheckingPointer) {
```
Comment (gysit) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:294]: This logic is quite hard to follow could it make sense to split in a part that checks gep.getIndices().front() and one that check gep.getIndices().drop_front()?

File: mlir/lib/Transforms/SROA.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+          DestructibleMemorySlot slot;
+          MemorySlotDestructionInfo info;
+        };
+
+        std::vector<DestructionJob> toDestruct;
+
+        for (Block &block : region)
+          for (Operation &op : block.getOperations())
+            if (auto allocator =
+                    llvm::dyn_cast<DestructibleAllocationOpInterface>(op))
```
Comment (gysit) [inline mlir/lib/Transforms/SROA.cpp:179]: usually we try to avoid too much nesting depth. Can we use `walk([&](DestructibleAllocationOpInterface allocator)`?

**Review by gysit — COMMENTED** [review]
Overall very nice!

I dropped a lot of comments and will need a second review to get more in depth with some parts of the algorithm. Especially some of the blocking uses handling was a bit confusing. And also the GEP stuff. I think it may make sense to add some more comments for the complicating things.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 5]
```diff
@@ -232,7 +233,9 @@ def LLVM_AllocaOp : LLVM_Op<"alloca",
 }
 
 def LLVM_GEPOp : LLVM_Op<"getelementptr", [Pure,
-    DeclareOpInterfaceMethods<PromotableOpInterface>]> {
+    DeclareOpInterfaceMethods<PromotableOpInterface>,
+    DeclareOpInterfaceMethods<TypeSafeOpInterface>,
+    DeclareOpInterfaceMethods<DestructibleAccessorOpInterface>]> {
```
[the remark this answers is not in the document]
Comment (gysit) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:238]: ok ✅ that sounds stupid (I mean that you need to implement both)

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.h [hunk 6]
```diff
@@ -23,6 +23,33 @@ struct MemorySlot {
...
+  /// Index of this memory slot in the parent memory slot.
+  Attribute subelementIndex;
+};
+
+struct DestructibleMemorySlot {
+  MemorySlot slot;
+  DestructibleSlotInfo info;
+};
+
+struct SubElementMemorySlot {
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.h:42]: This has been removed now, but I did not go with inheritance because that led to uncomfortable diamond-shaped inheritance graphs. Composition works best in that context.

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td [hunk 7]
```diff
@@ -186,11 +192,147 @@ def PromotableOpInterface : OpInterface<"PromotableOpInterface"> {
...
+    TODO
+  }];
+  let cppNamespace = "::mlir";
+
+  let methods = [
+    InterfaceMethod<[{
+        Returns whether all accesses in this operation to the provided slot are
+        done in a type-safe manner. To be type-safe, the access must only load
+        the value in this type as the type of the slot, and without assuming any
+        context around the slot. For example, a type-safe load must not load
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:256]: What's wrong with TypeSafeOpInterface?

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td [hunk 8]
```diff
@@ -186,11 +192,147 @@ def PromotableOpInterface : OpInterface<"PromotableOpInterface"> {
...
+  }];
+  let cppNamespace = "::mlir";
+
+  let methods = [
+    InterfaceMethod<[{
+        Destructs the type into subelements into a map of index attributes to
+        types of subelements. Returns nothing if the type cannot be destructed.
+      }],
+      "::llvm::Optional<::llvm::DenseMap<::mlir::Attribute, ::mlir::Type>>",
+      "destruct",
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:322]: I went with `getDestructedLayout`, tell me if you like it.

File: mlir/include/mlir/Transforms/Passes.td [hunk 9]
```diff
@@ -214,6 +214,13 @@ def SCCP : Pass<"sccp"> {
   let constructor = "mlir::createSCCPPass()";
 }
 
+def SROA : Pass<"sroa"> {
```
Comment (gysit) [inline mlir/include/mlir/Transforms/Passes.td:217]: This still needs some polishing :)

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+  auto numElem = dyn_cast<LLVM::ConstantOp>(getArraySize().getDefiningOp());
+  if (!numElem)
+    return {};
+  auto numElemIntAttr = numElem.getValue().dyn_cast<IntegerAttr>();
+  if (!numElemIntAttr || !numElemIntAttr.getType().isSignlessInteger() ||
+      numElemIntAttr.getInt() != 1)
+    return {};
+
+  Type elemType =
+      getElemType() ? *getElemType() : getResult().getType().getElementType();
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:67]: What I can do is use `has_value` on getElemType. That way, once it becomes not optional, this will be a compile error. What do you think?

File: mlir/include/mlir/Transforms/Passes.td [hunk 11]
```diff
@@ -214,6 +214,13 @@ def SCCP : Pass<"sccp"> {
...
+  let summary = "TODO";
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/Passes.td:218]: Does not look ready for review 😈

File: mlir/include/mlir/Transforms/Passes.td [hunk 12]
```diff
@@ -214,6 +214,13 @@ def SCCP : Pass<"sccp"> {
...
+  let description = [{
+    TODO
```
Comment (definelicht) [inline mlir/include/mlir/Transforms/Passes.td:220]: Does not look ready for review 😈

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp
Comment (definelicht) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:1]: How do I know what's new? 🥲

File: mlir/lib/Transforms/SROA.cpp [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+        continue;
+      }
+    }
+
+    // If it cannot be shown that the operation uses the slot safely, maybe it
+    // can be promoted out of using the slot?
+    scheduleAsBlockingUse(use);
+  }
+
+  SmallPtrSet<OpOperand *, 16> dealtWith;
```
Comment (definelicht) [inline mlir/lib/Transforms/SROA.cpp:53]: I would either use something more descriptive, or revert to `seen`. "Dealt with" sounds a bit informal 😁

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+  Type i32 = IntegerType::get(getContext(), 32);
+  DenseMap<Attribute, Type> destructured;
+  for (Type elemType : getBody()) {
+    destructured.insert({IntegerAttr::get(i32, index), elemType});
+    index++;
+  }
+  return destructured;
+}
+
+Type LLVM::LLVMStructType::getTypeAtIndex(Attribute index) {
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:394]: No! It is actually different. While the map is only generated if the type is reasonably small (because if your array is a million element wide that would create a map with a million elements), getTypeAtIndex can get a type from a destructible type no matter how big it is.

File: mlir/lib/Transforms/SROA.cpp [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+  while (!usedSafelyWorklist.empty()) {
+    MemorySlot mustBeUsedSafely = usedSafelyWorklist.pop_back_val();
+    for (OpOperand &subslotUse : mustBeUsedSafely.ptr.getUses()) {
+      if (dealtWith.contains(&subslotUse))
+        continue;
+      dealtWith.insert(&subslotUse);
```
Comment (definelicht) [inline mlir/lib/Transforms/SROA.cpp:59]:
```suggestion
      if (!dealtWith.insert(&subslotUse).second)
        continue;
```
🤓

File: mlir/lib/Transforms/SROA.cpp [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+                        DestructibleAllocationOpInterface allocator,
+                        OpBuilder &builder, MemorySlotDestructionInfo &info) {
+  OpBuilder::InsertionGuard guard(builder);
+
+  builder.setInsertionPointToStart(slot.ptr.getParentBlock());
+  DenseMap<Attribute, MemorySlot> subslots =
+      allocator.destruct(slot, info.usedIndices, builder);
+
+  llvm::SetVector<Operation *> usersToRewire;
+  for (auto &[user, _] : info.userToBlockingUses)
```
Comment (definelicht) [inline mlir/lib/Transforms/SROA.cpp:117]:
```suggestion
  for (auto &[user, _] : info.userToBlockingUses)
```
I don't think `_` means an unnamed variable in C++, you're binding it to a variable called `_`. I would either admit that you're naming it, or fall back on `it.second`

File: mlir/lib/Transforms/SROA.cpp [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+    usersToRewire.insert(user);
+  for (DestructibleAccessorOpInterface accessor : info.accessors)
+    usersToRewire.insert(accessor);
+  SetVector<Operation *> sortedUsersToRewire =
+      mlir::topologicalSort(usersToRewire);
```
Comment (definelicht) [inline mlir/lib/Transforms/SROA.cpp:122]: Could it make sense to reassign `usersToRewire` to its sorted equivalent to not allocate yet another set vector in this (humongous) scope?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+    if (isCheckingPointer) {
+      isCheckingPointer = false;
+      if (indexInt.getInt() != 0)
+        return {};
+      continue;
+    }
+    assert(!selectedType.isa<LLVM::LLVMPointerType>());
+    if (!firstLevelIndex)
+      firstLevelIndex = indexInt;
+    auto destructible = selectedType.dyn_cast<DestructibleTypeInterface>();
```
[the remark this answers is not in the document]
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:306]: I assume I missed that selected type is updated in the loop. Still think some improvements to this loop nest or at least some commenting what is going on could be very helpful.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 19]
```diff
@@ -0,0 +1,419 @@
...
+                            SmallVectorImpl<MemorySlot> &mustBeSafelyUsed) {
+  if (getBase() != slot.slot.ptr || slot.slot.elemType != getElemType())
+    return false;
+  auto [reachedType, firstLevelIndex] = computeReachedGEPType(*this);
+  if (!reachedType || !firstLevelIndex)
+    return false;
+  assert(slot.info.elementsPtrs.contains(firstLevelIndex));
+  if (!slot.info.elementsPtrs.at(firstLevelIndex).isa<LLVM::LLVMPointerType>())
+    return false;
+  mustBeSafelyUsed.emplace_back<MemorySlot>({getResult(), reachedType});
```
[the remark this answers is not in the document]
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:343]: I don't remember what my though was here... not sure if comments can shift in github but there is no check here...

File: mlir/lib/Transforms/SROA.cpp (hunk 16 above)
Comment (gysit) [reply mlir/lib/Transforms/SROA.cpp:117]: there was this make first range thing that Jakub suggested?

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 20]
```diff
@@ -0,0 +1,429 @@
...
+  int32_t indexInt = indexAttr.getInt();
+  ArrayRef<Type> body = getBody();
+  if (indexInt < 0 || body.size() <= ((uint32_t)indexInt))
+    return {};
+  return body[indexInt];
+}
+
+Optional<DenseMap<Attribute, Type>>
+LLVM::LLVMArrayType::getDestructedLayout() const {
+  if (getNumElements() > 16)
```
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:410]: NIT: Add a constant expression somewhere to not have an arbitrary constant in the code here.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 21]
```diff
@@ -0,0 +1,429 @@
...
+    return {};
+  int32_t numElements = getNumElements();
+
+  Type i32 = IntegerType::get(getContext(), 32);
+  DenseMap<Attribute, Type> destructured;
+  for (int32_t index = 0; index < numElements; index++)
```
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:416]: Omega ultra nit: 
```suggestion
  for (int32_t index = 0; index < numElements; ++index)
```

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp [hunk 22]
```diff
@@ -0,0 +1,429 @@
...
+    destructured.insert({IntegerAttr::get(i32, index), getElementType()});
+  return destructured;
+}
+
+Type LLVM::LLVMArrayType::getTypeAtIndex(Attribute index) const {
+  auto indexAttr = index.dyn_cast<IntegerAttr>();
+  if (!indexAttr || !indexAttr.getType().isInteger(32))
+    return {};
+  int32_t indexInt = indexAttr.getInt();
+  if (indexInt < 0 || getNumElements() <= ((uint32_t)indexInt))
```
Comment (Dinistro) [inline mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:426]: `static_cast`

**Review by Dinistro — COMMENTED** [review]
Added some NIT comments.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp]: It's used once in mem2reg and once in SROA.
What I meant is that with the explicit use, whenever the optional disappear this will create a compile error, calling for the programmer to remove this now unnecessary code. I can make a local static helper though, this should be a good compromise.

[the remark this answers is not in the document]
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp]: The current code will already produce a compile error since getElemType() will return the type instead of an optional containing the type. I think you can just mark this comment as resolved then!

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:397]: No no. `getTypeAtIndex` is for example used when you want to ensure the type safety of further uses. Say you have
`struct { i32, i32, array<10000000000 x i32> }`. It is definitely reasonable to SROA this at the first order. So you would ask for its destruction, leading to three new allocas: `i32`, `i32` and `array<10000000000 x i32>`. But you still need to check that those new allocas would be used in a type-safe manner. So if the last alloca had a further gep that would access the content of the array, you need to check that this access is type-safe. But of course it would be unreasonable to destruct array<10000000000 x i32>`. This is when getTypeAtIndex is useful: it will give you information about large aggregate types without having to generate the huge maps that destructing them would. This can be used even if the destruction request is not possible.

File: mlir/include/mlir/Transforms/Passes.td (hunk 9 above)
Comment (Moxinilian) [reply mlir/include/mlir/Transforms/Passes.td:217]: Tell me if what I wrote is satisfying.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (gysit) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:405]: Ok I understand. Are we checking the array accesses to make sure they are not out-of-bounds? Type safety on it's own should not be a concern as long as all accesses are inbounds? Or am I missing something?

File: mlir/include/mlir/Transforms/Passes.td (hunk 9 above)
Comment (gysit) [reply mlir/include/mlir/Transforms/Passes.td:217]: Sounds good! Should we also add a pass flag that limits the size of the aggregates we consider for destruction (probably in terms of number of elements?).

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.h
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.h:1]: Yes, SROA is not considered part of mem2reg. Even though both are often used together, they do very independent things.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h [hunk 23]
```diff
@@ -198,6 +200,12 @@ class LLVMStructType
 
   LogicalResult verifyEntries(DataLayoutEntryListRef entries,
                               Location loc) const;
+
+  /// Destructs the struct into its indexed field types.
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h:204]: Let's go for destructure.

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:1]: So it's not super clear where the boundary is. There is "PromotableOpInterface" which is used by both, "TypeSafeOpInterface" which is used by SROA but has nothing SROA-specific to it. Generally I think those interfaces are meant to be used in any setting and happen to be used by mem2reg and SROA.

File: mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp
Comment (Moxinilian) [reply mlir/lib/Dialect/LLVMIR/IR/LLVMMemorySlot.cpp:1]: Blame Torvalds!

File: mlir/include/mlir/Transforms/Mem2Reg.h [hunk 24]
```diff
@@ -11,10 +11,111 @@
...
 
+/// Information computed during promotion analysis used to perform actual
+/// promotion.
+struct MemorySlotPromotionInfo {
+  /// Blocks for which at least two definitions of the slot values clash.
+  SmallPtrSet<Block *, 8> mergePoints;
+  /// Contains, for each operation, which uses must be eliminated by promotion.
+  /// This is a DAG structure because an operation that must eliminate some of
+  /// its uses always comes from a request from an operation that must
+  /// eliminate some of its own uses.
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Transforms/Mem2Reg.h:26]: Tell me if you prefer the new one.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h [hunk 25]
```diff
@@ -198,6 +200,12 @@ class LLVMStructType
...
+  Optional<DenseMap<Attribute, Type>> getDestructedLayout();
+
+  /// Returns which type is stored at a given integer index within the struct.
+  Type getTypeAtIndex(Attribute index);
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMTypes.h:208]: So the reason for this is that sometimes indices are not integers but fancy things. For example, if you wanted to use SROA on for example a hashmap dialect, your indices could be anything.

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td [hunk 26]
```diff
@@ -186,11 +189,148 @@ def PromotableOpInterface : OpInterface<"PromotableOpInterface"> {
...
+      "::llvm::SmallVector<::mlir::DestructibleMemorySlot>",
+      "getDestructibleSlots",
+      (ins)
+    >,
+    InterfaceMethod<[{
+        Destructures this slot into multiple subslots. The newly generated slots
+        may belong to a different allocator. The original slot must still exist
+        at the end of this call.
+
+        The builder is located at the beginning of the block where the slot
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:223]: I agree. This comes from suggestions upstream, though. I don't know if I should try to fight it too much.

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td [hunk 27]
```diff
@@ -186,11 +189,148 @@ def PromotableOpInterface : OpInterface<"PromotableOpInterface"> {
...
+    Describes operations allocating memory slots of aggregates that can be
+    destructed into multiple smaller allocations.
+  }];
+  let cppNamespace = "::mlir";
+
+  let methods = [
+    InterfaceMethod<[{
+        Returns the list of slots for which destruction should be attempted,
+        specifying in which way the slot should be destructed into subslots. The
+        subslots are indexed by attributes. This computes the type of the
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:210]: Well no, because of what I mentioned above. I don't really know how to word it otherwise :/

File: mlir/include/mlir/Interfaces/MemorySlotInterfaces.td [hunk 28]
```diff
@@ -186,11 +189,148 @@ def PromotableOpInterface : OpInterface<"PromotableOpInterface"> {
...
+        must implement `DestructibleTypeInterface`.
+      }],
+      "::llvm::SmallVector<::mlir::DestructibleMemorySlot>",
+      "getDestructibleSlots",
+      (ins)
+    >,
+    InterfaceMethod<[{
+        Destructures this slot into multiple subslots. The newly generated slots
+        may belong to a different allocator. The original slot must still exist
+        at the end of this call.
```
[the remark this answers is not in the document]
Comment (Moxinilian) [reply mlir/include/mlir/Interfaces/MemorySlotInterfaces.td:221]: `destruct` destructs the slot but does not destroy it yes :)
I'm changing the name for `destructure` to make Johannes happy.

**Comment by Moxinilian** [comment]
I'm going to squash all this and manually separate it into two different PRs. Feel free to finish the ongoing discussions here and move further comments to the new PRs once they are ready.
